### PR TITLE
[MM-36406] Apps support for Plugins

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
To support Apps run by plugin this PR
- Adds REST API methods to instal/uninstall/get/disable/enable an app
- Adds a plugin upstream that makes use of the inter-plugin communication. This includes a `PluginID` field in `App`. All outgoing request go to `/$PLUGIN_ID/app`.
- Combines the RPC driver and the HTTP driver into a `MMClient` interface that `Proxy` uses to make API call on the MM sever
- Move authentication from  proxy layer to http layer to allow checking based on the HTTP request issuer (plugin or user)

Please leave plenty of feedback.

The commit history is not clean, but maybe helps with reviewing the changes.

The PR targets https://github.com/mattermost/mattermost-plugin-apps/pull/207 as it makes use of enable/disable.

#### TODO
- [x] Fix tests
- [ ] Update E2E build image
- [ ] Add tests
- [ ] Write down use cases
- [x] Add uninstall endpoint
- [x] Hide apps run by plugins in the Marketplace
- [x] Hide apps run by plugins in `/apps list`

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-36407
https://mattermost.atlassian.net/browse/MM-36408
https://mattermost.atlassian.net/browse/MM-36409

### CODE REVIEW

File: server/proxy/enable.go
Comment (larkox) [inline server/proxy/enable.go:25]: Are we creating a new bot for plugin apps? I guess we shouldn't, but not 100% sure.

File: server/proxy/install.go
Comment (larkox) [inline server/proxy/install.go:149]: Is the function objective to create or to ensure the bot?

**Review by larkox — COMMENTED** [review]
Awesome work hanzei. Several comments here and there.

File: .golangci.yml
Comment (hanzei) [reply .golangci.yml:16]: The linter suggests reordering of fields in structs to save memory. While there are a few place where this sense as they are used in hot areas (e.g. `CallRequest`) it overkill for the whole code base. Please note that the starter template also disabled it: https://github.com/mattermost/mattermost-plugin-starter-template/commit/f5cae51a20a81dd87e11a793a3bb494d6800c9c0

File: Makefile
Comment (hanzei) [reply Makefile:100]: Good question. Changed it to a fixed version to reduce the chance of build issues.

File: apps/app.go
Comment (hanzei) [reply apps/app.go]: Fixed in c769123

File: apps/mmclient/mattermost_client_pp.go
Comment (hanzei) [reply apps/mmclient/mattermost_client_pp.go:64]: Maybe. I would like to come back to this code part and fix the naming and usability issues (remove `model.AppError´), but the PR is already quite big.

Would you be fine deferring this change to a new ticket?

File: server/httpin/restapi/app.go
Comment (hanzei) [reply server/httpin/restapi/app.go:25]: 1. A plugin has root access to the system and hence should be able to get information on any app.
2. Yes, I think we should keep that possibility open. The current design allows that if the apps routes requests on an application level. (All call go to `/app/`)

File: server/httpin/restapi/restapi.go
Comment (hanzei) [reply server/httpin/restapi/restapi.go:69]: I've renamed the middle ware to `CheckPluginOrUser` to make clear that it only checks _who_ made the request. The handler still has to check the permissions.

File: upstream/uphttp/static.go
Comment (hanzei) [reply upstream/uphttp/static.go]: Good suggestion. Added in eca2650.

File: server/proxy/enable.go
Comment (hanzei) [reply server/proxy/enable.go:25]: Well, the bot name is based on the `AppID`. If the plugin wants to use the same bot account, it needs to use an appropriate ID.

File: server/proxy/install.go
Comment (hanzei) [reply server/proxy/install.go:149]: To create. It should not exists, given that https://github.com/mattermost/mattermost-plugin-apps/blob/20597e47b253887e53616fb5fe83ab063b7a6a5c/server/proxy/install.go#L147-L147 returned `nil`.

**Comment by hanzei** [comment]
https://github.com/mattermost/mattermost-server/pull/17832 need to be merged I can update the docker tag used for the E2E test to fix them.

File: apps/mmclient/mattermost_client_pp.go
Comment (larkox) [reply apps/mmclient/mattermost_client_pp.go:65]: Sure, no problem from my side.

File: server/httpin/restapi/app.go
Comment (larkox) [inline server/httpin/restapi/app.go:187]: PRC or RPC?

**Review by larkox — APPROVED** [review]
LGTM, just a small typo, I think.

File: server/httpin/restapi/app.go
Comment (hanzei) [reply server/httpin/restapi/app.go]: `RPC`. Good catch :+1: Fixed in 7527b1.

**Comment by hanzei** [comment]
Heads up that I moved the MMClient into a separate package in 05e366a61f9f11cc57402c6ca65bb8f4c8db2378.

File: apps/mmclient/mattermost_client_pp.go
Comment (hanzei) [reply apps/mmclient/mattermost_client_pp.go:65]: Created https://mattermost.atlassian.net/browse/MM-36819

File: server/proxy/enable.go
Comment (mickmister) [reply server/proxy/enable.go:26]: Does this mean a plugin's app id must be equal to its existing bot's username?

Comment (hanzei) [reply server/proxy/enable.go:26]: > Does this mean a plugin's app id must be equal to its existing bot's username

If it wants to use the bot, yes. 

I'm hesitant to introduce more complexity by adding bonus features just for apps run by plugins.

File: mmclient/rpc.go
Comment (catalintomai) [inline mmclient/rpc.go:44]: typo

Comment (hanzei) [reply mmclient/rpc.go]: Fixed in d563526

File: apps/mmclient/mattermost_client_pp.go
Comment (levb) [inline apps/mmclient/mattermost_client_pp.go:62]: I don't see `fromPlugin` being set anywhere - is it needed?

File: utils/httputils/utils.go
Comment (hanzei) [reply utils/httputils/utils.go:129]: It's unneeded. Removed the method in bd19dc1.

File: server/proxy/uninstall.go
Comment (hanzei) [reply server/proxy/uninstall.go:16]: I'm 2/5 that the proxy should not care about which client is uses. In the future it might not even be an admin one.

File: apps/mmclient/mattermost_client_pp.go
Comment (hanzei) [reply apps/mmclient/mattermost_client_pp.go:62]: It's used in https://github.com/mattermost/mattermost-plugin-apps/blob/722b4066c8b9b27d1fab06239eca4475b8ca495b/apps/mmclient/mattermost_client_pp.go#L239-L241

to build the base URL.

Comment (hanzei) [reply apps/mmclient/mattermost_client_pp.go:62]: It's used in https://github.com/mattermost/mattermost-plugin-apps/blob/722b4066c8b9b27d1fab06239eca4475b8ca495b/apps/mmclient/mattermost_client_pp.go#L237-L244
to build the base URL.

File: apps/mmclient/mattermost_client.go
Comment (hanzei) [reply apps/mmclient/mattermost_client.go:38]: The whole rest client(s) need an overhaul of the API. (See also my long post in the apps channel) I'm inclined to deffer the naming discussion to that effort.

About your suggestion: `NewAppsClient` sounds reasonable to me, but there is already a `NewClient` method in `mattermost_client.go`.

0/5 revert to `NewAPIClientPP`?

File: utils/httputils/utils.go
Comment (hanzei) [reply utils/httputils/utils.go]: It's unneeded. Removed the method in bd19dc1.

File: utils/httputils/utils.go
Comment (hanzei) [reply utils/httputils/utils.go]: bc8941d

File: server/proxy/proxy.go
Comment (hanzei) [reply server/proxy/proxy.go]: :+1: I've also renamed `getAsset` to `getStatic`. See b78c4469a17432d96788821aad59728de4ee4df7.

**Review by levb — APPROVED** [review]
I may submit a few proposed (cosmetic) changes in a separate PR.

**Comment by hanzei** [comment]
@levb Would you please merge this PR after QA is complete?

**Comment by DHaussermann** [comment]
I have done a round of testing on auto-link https://github.com/mattermost/mattermost-plugin-autolink/pull/164

The autolink app does work and no regressions were found when testing the install flow for conventional apps. However, I do see a couple issues.
1. On install (as well as subsequent disable / enables) the binding does not appear until I switch to a new channel
2. The app does not get included in the `/apps list`
3. If I remove Auto-link and reinstall it enabling the experimental shows the following error. 
image

For 1. and 2. I see the following in the logs:
```
INFO[2021-07-20T11:38:29.785964-04:00] Installed an app                              app_id=dylan caller="mlog/sugar.go:21" plugin_id=com.mattermost.apps
DEBU[2021-07-20T11:38:29.788525-04:00] Not auto installing/upgrade because plugin was disabled  caller="mlog/log.go:230" plugin_id=com.mattermost.plugin-incident-management version=1.14.3
WARN[2021-07-20T11:38:29.798879-04:00] Failed to fetch license two times. Falling back to on-prem mode.  caller="mlog/sugar.go:29" plugin_id=com.mattermost.apps
DEBU[2021-07-20T11:38:29.8281322-04:00] Schedulers received config change.            caller="jobs/schedulers.go:234"
DEBU[2021-07-20T11:38:29.8394089-04:00] Next run time for scheduler                   caller="jobs/schedulers.go:216" next_runtime="<nil>" scheduler_name=MigrationsScheduler
DEBU[2021-07-20T11:38:29.849737-04:00] Next run time for scheduler                   caller="jobs/schedulers.go:216" next_runtime="2021-07-21 11:38:29.849715 -0400 EDT m=+87859.579334406" scheduler_name=PluginsScheduler
DEBU[2021-07-20T11:38:29.8530612-04:00] Received HTTP request                         caller="web/handlers.go:105" method=POST request_id=yhogremd6pfs8b3d7gymhponno status_code=200 url=/api/v4/plugins/dylan/enable
DEBU[2021-07-20T11:38:29.8611782-04:00] Next run time for scheduler                   caller="jobs/schedulers.go:216" next_runtime="2021-07-20 12:38:29.86117 -0400 EDT m=+5059.590788678" scheduler_name=ProductNoticesScheduler
DEBU[2021-07-20T11:38:29.873268-04:00] Next run time for scheduler                   caller="jobs/schedulers.go:216" next_runtime="2021-07-20 11:38:34.873254 -0400 EDT m=+1464.602872959" scheduler_name=ResendInvitationEmailJobScheduler
DEBU[2021-07-20T11:38:29.8811939-04:00] Next run time for scheduler                   caller="jobs/schedulers.go:216" next_runtime="2021-07-21 11:38:29.881187 -0400 EDT m=+87859.610805475" scheduler_name=ImportDeleteScheduler
DEBU[2021-07-20T11:38:29.888247-04:00] Next run time for scheduler                   caller="jobs/schedulers.go:216" next_runtime="2021-07-21 11:38:29.888238 -0400 EDT m=+87859.617856532" scheduler_name=ExportDeleteScheduler
DEBU[2021-07-20T11:38:37.112302-04:00] plugin process exited                         caller="plugin/hclog_adapter.go:52" plugin_id=mattermost-autolink wrapped_extras=pathplugins/mattermost-autolink/server/dist/plugin-darwin-amd64pid63281
DEBU[2021-07-20T11:38:37.1123352-04:00] plugin exited                                 caller="plugin/hclog_adapter.go:54" plugin_id=mattermost-autolink
```
```
ERRO[2021-07-20T12:06:23.022391-04:00] Access to route for non-existent plugin in inter plugin request  caller="mlog/log.go:251" destination_plugin_id=mattermost-autolink error="plugin not found: mattermost-autolink" source_plugin_id=com.mattermost.apps url=/app/bindings
not a valid logrus Level: "dpanic"
```
```
ERRO[2021-07-20T12:06:40.580418-04:00] Access to route for non-existent plugin in inter plugin request  caller="mlog/log.go:251" destination_plugin_id=mattermost-autolink error="plugin not found: mattermost-autolink" source_plugin_id=com.mattermost.apps url=/app/bindings
DEBU[2021-07-20T12:06:40.581402-04:00] Received HTTP request                         caller="web/handlers.go:105" method=GET request_id=memtjis7uiff5j35epms1h4cmo status_code=200 url=/api/v4/channels/gpunn98w8tngjbkhireayd7qjw/posts
not a valid logrus Level: "dpanic"
```

**Comment by DHaussermann** [comment]
@levb I'm curious if this PR is blocking for the work you have ongoing. Not sure if the issues above should be investigated now or addressed separately after this is merged?

**Comment by levb** [comment]
@DHaussermann The issues in Autolink are specific to the implementation in the (autolink) plugin itself, and I am proposing that we do NOT merge the autolink changes, yet. The behavior of the Apps plugin appears "as expected".

I am going to resolve the conflicts now so we can wrap this up?

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
- Tested with Autolink as well as a 2nd app provided by @mickmister 
- Plugins expose the app as expected
- The app is functional
- No regression found for install and uninstall of plugins and conventional apps
LGTM!

Documented auto-link issues here https://github.com/mattermost/mattermost-plugin-autolink/pull/164

**Comment by DHaussermann** [comment]
Sounds good @levb Please merge.

**Comment by levb** [comment]
@DHaussermann I just merged the master changes, there were material conflicts. Do you mind a quick regression pass (deploy/use autolink and http)?

**Comment by DHaussermann** [comment]
@levb  retesting looks pretty good. No new issues were found. The issue were the "Plugins-Apps" does not get included in the _/apps list_ is unfortunately still present. Maybe this is only working intermittently?

I'd be okay merging and creating a separate issue to address this.

**Comment by levb** [comment]
`/apps list --plugin-apps` I think lists it

**Pull request merged**
